### ISSUE TITLE
Official Xposed for N is incompatible with Samsung Near-Stock ROM

### ISSUE THREAD

**Issue by svg-frog** [rovo89/XposedBridge#192]
I sort of expected this to be honest, but since (I'm assuming) you worked with a OP5 for N, linked is the startup logcat from it bootlooping. Seems like dex2oat doesn't have the correct arguments, and the system isn't able to startup regardless.

Sorry about how long the file is, I let it go for a while to see if it was just taking a while to recompile Xposed. Most of the stuff is just repeating retries though.

Edit: This is from a Galaxy S8+ with the Exynos SoC. Specifically, the custom ROM is Rennovate.

[[Pastebin](https://pastebin.com/raw/CC0tLtQz)]

**Comment by rovo89** [comment]
Thanks for creating the bug report and including the log! That's something I can work with, and more is actually more here. :smile:

```
10-08 13:52:15.665  1024  1024 E dex2oat : Unexpected CPU variant for Arm64: exynos-m2
```

Please try this: https://www.dropbox.com/s/tzlht5far8gw5pz/xposed-v89-sdk24-arm64-test1.zip?dl=0

**Comment by OrpheeGT** [comment]
Just tested your test1 zip ! Boots fine ! Great !

**Comment by wanam** [comment]
Nice catch! I guess the fix relay here:
https://github.com/rovo89/android_art/blob/xposed-nougat-mr2/runtime/arch/arm64/instruction_set_features_arm64.cc#L42

**Comment by Firefds** [comment]
the 89-test1 zip now works great on the fully stock G950F S8, thanks rovo89!

**Comment by svg-frog** [comment]
@rovo89 Can confirm, Xposed-89-Test1 boots successfully on my S8+. Thank you for your work on Xposed, and the quick response to this issue.

**Comment by AngelIsL33T** [comment]
Tried this on S8 SD. Didn't work, Hopefully you can get the SD variant fixed soon. Great job btw!

**Comment by svg-frog** [comment]
@AngelIsL33T If you want to help, connect your phone to your PC, and with Android Debug Bridge, in a command prompt, type:
adb logcat > startup.txt

Then after its been stuck booting for ~>5 minutes, upload the log to pastebin and attach it to here. Without a log its kinda hard for a developer to do anything.

**Comment by Ethaligan** [comment]
Changed 'dalvik.vm.isa.arm64.variant' build.prop key from 'exynos-m2' to 'exynos-m1' and it seems to work fine.
No more error messages.

Currently on S8+ (SM-G955F) running 7.1.1

**Comment by svg-frog** [comment]
@Ethaligan If you're on 7.1.1, yeah, don't use the test zip attached above. Thats for sdk24 and you're on sdk25.

**Comment by koko115** [comment]
Test1 works me to on S8 plus 7.0 deodex

Now how make xposed with CTS pass ;-)

**Comment by HorstiG** [comment]
I'm on S8 Plus with Note 8 ROM Port, SDK25 is active, but works very slowly, the RAM consumption has increased by half. I have only activated one module, Flat Sytle Bar Indicators. S7 Edge with N 7.0 unofficial Xposed version works perfectly.

**Comment by svg-frog** [comment]
@HorstiG If you just installed, Android is now rebuilding the cache to speed up apps. Give it a bit.

**Comment by HorstiG** [comment]
I forgot, after each reboot, Android will update 166 apps, like after cleaning Dalvik and Cache. 
Working with Xposed is not possible at the moment, even after 2 hours

**Comment by Pkhrl** [comment]
Can anyone Explain that one error please. It is slowing down my device and causing apps to randomly crash.
Details:  s6 Edge, sm-g925i , running android 7.0, magisk v14, xposed v89test1,

Xposed log:

Added Xposed (/system/framework/XposedBridge.jar) to CLASSPATH
10-09 11:59:18.534 E/Xposed (25108): Failed to open oat /system/framework/oat/arm64/services.odex: File /system/framework/oat/arm64/services.odex is too short to be a valid ELF file
10-09 11:59:18.719 I/Xposed (25108): Detected ART runtime
10-09 11:59:18.720 I/Xposed (25108): Found Xposed class 'de/robv/android/xposed/XposedBridge', now initializing
10-09 11:59:20.505 I/Xposed (25117): -----------------
10-09 11:59:20.505 I/Xposed (25117): Starting Xposed version 88, compiled for SDK 24
10-09 11:59:20.505 I/Xposed (25117): Device: Pixel 2 XL (Google), Android version 7.0 (SDK 24)
10-09 11:59:20.505 I/Xposed (25117): ROM: NRD90M.G925IDVU3FQD1
10-09 11:59:20.505 I/Xposed (25117): Build fingerprint: samsung/zeroltedv/zerolte:7.0/NRD90M/G925IDVU3FQD1:user/release-keys
10-09 11:59:20.505 I/Xposed (25117): Platform: arm64-v8a, 32-bit binary, system server: no
10-09 11:59:20.505 I/Xposed (25117): SELinux enabled: yes, enforcing: yes
10-09 11:59:20.520 I/Xposed (25117): -----------------

Thanks in Advance.

**Comment by AngelIsL33T** [comment]
@ElectronicWizard when would I do this? Give instructions please. Do I type that while stuck in bootloop?

**Comment by svg-frog** [comment]
@AngelIsL33T Yes. If its been in a bootloop a while, just type the command, wait about 5 seconds, then hit Ctrl + C, then upload the resulting file to a website like Pastebin.com.

**Comment by AngelIsL33T** [comment]
@ElectronicWizard alright bro. So adb is active while on boot screen?

**Comment by svg-frog** [comment]
@AngelIsL33T I don't know if *all* functions are, but adb logcat is enabled, yes.

**Comment by AngelIsL33T** [comment]
Ok. will try it now. should I flash xposed thru flashfire, or thru the Apk?

**Comment by svg-frog** [comment]
@AngelIsL33T I was flashing using TWRP. If you're using that test ZIP from above you will need to flash through recovery.

**Comment by AngelIsL33T** [comment]
@ElectronicWizard I'm using SamPWND root. If that matters or not.

**Comment by rovo89** [comment]
> Can confirm, Xposed-89-Test1 boots successfully on my S8+.

Great to hear!


> I'm on S8 Plus with Note 8 ROM Port, SDK25 is active, but works very slowly, the RAM consumption has increased by half. I have only activated one module, Flat Sytle Bar Indicators. S7 Edge with N 7.0 unofficial Xposed version works perfectly.

> I forgot, after each reboot, Android will update 166 apps, like after cleaning Dalvik and Cache.

@HorstiG Please post the full logs. Maybe something is going wrong during compilation. Also, do you see this as a popup? Because yes, Android will recompile some apps, but that should happen in the background. It would simply interpret+JIT the apps you use before they're compiled.

> 10-09 11:59:18.534 E/Xposed (25108): Failed to open oat /system/framework/oat/arm64/services.odex: File /system/framework/oat/arm64/services.odex is too short to be a valid ELF file

@Pkhrl Please upload that file, maybe it's corrupted. Other possibility would be that the process isn't allowed to read it.

**Comment by Pkhrl** [comment]
https://drive.google.com/file/d/0B1RhXBqL3xCqclo0X1EtOTZ5Rzg/view?usp=drivesdk

It must be corrupted. It's 0 Byte . I don't know. May be it's causing to slow down my device as it keep failing continuously. Thanks for responding Rovo

**Comment by rovo89** [comment]
@Pkhrl Yeah, a 0 byte file is definitly wrong. Could you please help me to analyze this? Are there other .odex files which are 0 bytes? Does reflashing the system image help? Any other idea why the file might be destroyed? Usually, /system is read-only, so how could that happen? How did you install Xposed?

**Comment by Pkhrl** [comment]
I installed xposed through Twrp recovery this morning. First I installed xposed88 without bootloops on my samsung s6edge.I saw that error in xposed log and causing slow performance (obviously it is after installing xposed), then I installed v88test1 to see if this fixes issue. Still same, and apps like gallery, playstore, viber keeps crashing randomly, not when I run them, it's just random and Haven't affected my daily use.

Only services.odex file is 0 byte. I checked all files inside framework folder, none are 0 byte except services.odex .

I just noticed that arm64 folders inside framework folder are created after installing Xposed , is that right? 

Few weeks ago I deodexed services.jar to create smali patcher module for magisk, which generally treat fake location as real and allows to screenshot in any app.

**Comment by Pkhrl** [comment]
Is it normal? I have been getting same log over and over within few seconds. Thank you very much for responding.

**Comment by rovo89** [comment]
> I just noticed that arm64 folders inside framework folder are created after installing Xposed , is that right?

No, Xposed shouldn't touch anything in /system, apart from the installation of course. So you're saying that arm64 where created? Where exactly and what's inside?

> Few weeks ago I deodexed services.jar to create smali patcher module for magisk, which generally treat fake location as real and allows to screenshot in any app.

If you deodexed, then why is there an .odex file? All you should need is the services.jar with classes.dex inside. Maybe the timestamp shows that this empty file was left from the deodexing, so please delete that file and see if that helps.

**Comment by wanam** [comment]
@pkhrl what tool did you use for deodexing?

**Comment by Pkhrl** [comment]
I used FulmicsDeodexer_1.6 and followed this instructions https://forum.xda-developers.com/apps/magisk/module-smali-patcher-0-7-t3680053/amp/

**Comment by Pkhrl** [comment]
This was before installing Xposed. Is there any way that could speed up my system as usual? Obviously without installing Xposed

**Comment by Pkhrl** [comment]
When I try to delete system/framework/oat/arm64/services.odex it says read only file.  There are other odex file as well. Here are some screenshots of files inside framework folder.   https://drive.google.com/folderview?id=0B1RhXBqL3xCqV0k1VWNXeFlqYlU
Thanks again for analysing this issue which seems more due to magisk module but my system wasn't slow and crashing apps, so there might be some connection with Xposed. Hope you can help me out. thanks Robo and wanam.

**Comment by Pkhrl** [comment]
Here is some more screenshot from my laptop.
https://drive.google.com/folderview?id=0B1RhXBqL3xCqbDQzWDliekd0SXc

I don't know it was before or after deodoxing, but here is the system/framework folder from my device copied to my pc to deodox. I can't see any arm64,arm, oat folder. May be it was created after flashing smali module, which basically was created through services.jar.

**Comment by wanam** [comment]
How did you push the deodexed files? you shouldn't have the arm64/arm folders on your device.

**Comment by Pkhrl** [comment]
I'm not that technical, I just followed steps from https://forum.xda-developers.com/apps/magisk/module-smali-patcher-0-7-t3680053/amp/ and flashed smali module. Honestly I'm not sure.

**Comment by svg-frog** [comment]
@rovo89 has the exynos-m2 patch been pushed/released to the main download site, particularly for sdk25? Had someone asking about it for the Note 8/7.1.1.

**Comment by HorstiG** [comment]
Yes, for Note 8 Port on S8 Plus dream2lte. I have no Logfile sorry. No PC nearby.

**Comment by Jeffdow85** [comment]
Thanks @electronicwizard for bringing up my issue, I  have subscribed to this thread now to follow updates.

**Comment by rovo89** [comment]
@ElectronicWizard @Jeffdow85 This and other patches are part of v88.1, please give it a try.
@HorstiG Please also try this, maybe it helps. If not, I'll need the logs.
@Pkhrl Please try again to delete the 0-byte `services.odex` file. I personally use Root Explorer, which I think can also take care of remounting `/system` read-write. There are probably other apps as well which can do the job.

**Comment by adrienbrignon** [comment]
Just tried v88.1, working very well on Galaxy S8+ (SM-G955F) SDK 24.
Thanks @rovo89

**Comment by HorstiG** [comment]
V88.1 works so far, very long boot time and in the running operation not to use. I get out of here because I can not help without PC unfortunately. Thanks for the efforts. S8 dream2lte with Note 8 ROM Port. SDK25

**Comment by Jeffdow85** [comment]
Thanks @rovo89 will have to wait now as I'm locked out of flashing anything on the note 8 some stupid 7 day bootloader unlock Samsung seems to have implemented. Can't wait to try though thanks so much.

**Comment by Jasi2169** [comment]
@rovo89 rovo i can successfully boot and xposed is active in my s7 but it does not change any value in my module for example hooking system properties to change device name like it used to on marshmallow and below and there is no error in the logs

is there something i need to add to support modules for nougat+ ?

@wanam the old framework i installed on my s7 6.0.1 was wanam_unofficial build but i see rovo has updated new version of xposed and i believe it supports your patch/or fix bootlops for samsung stock roms like your unofficial build right ?

**Comment by rovo89** [comment]
> is there something i need to add to support modules for nougat+ ?

Nothing specific. You only need to change your module if whatever you're hooking has changed. For example, if you hook a system method in the Android framework itself and that method has been renamed between Android 6.0 and 7.0, you'll obviously need to adjust your findAndHookMethod() call to use the new name (or even better, decide which name to use based on Build.VERSION.SDK_INT). On the other hand, if you're hooking a user-installed app and that app is unchanged (maybe even the same APK is used), then the module just keeps working. The only thing you might need to consider is File-Based Encryption. If you're using shared preferences, those might not be readable in e.g. initZygote() now.

So much to the general question. If you still think there might be a bug in the framework, please open an issue and more details (e.g. the relevant parts of your code, error logs if any, smali code of the method you want to hook, ...).

**Comment by C3C0** [comment]
What if I need SharedPreferences within initZygote on devices with FBE? Is there a recommendation how to work around it? //sorry for OT

**Comment by rovo89** [comment]
I think the answer to that is https://developer.android.com/reference/android/content/Context.html#createDeviceProtectedStorageContext()

It uses `/data/user_de/0/<appid>` in the background for calls like `getFiles()`. I haven't added a special API to `XSharedPreferences` yet which uses this path instead of `/data/data/<appid>`, but you can use the full path to open your preferences file.

The problem seems to be that this directory doesn't exist on some ROMs, and I'm not sure why. I think it should be available everywhere, just not with a special meaning on unencrypted devices. Maybe the directory is created when we access it with official APIs like the one above, not sure.

**Comment by roberto-sartori-gl** [comment]
https://github.com/googlesamples/android-DirectBoot/blob/master/README.md

I'm not sure about this, but:
'Apps can mark individual components as being direct boot aware which indicates to the system that they can safely run when Credential protected storage is unavailable (an direct boot aware component primarily relies on data stored in the new Device protected storage area, but they may access Credential protected data when unlocked) by adding directBootAware="true" in the manifest.'

**Comment by Jasi2169** [comment]
the issue is with the preferences reading that is the reason why so many modules are not working on nougat,the function name is same but the sharedpreferences is having problem because the data partition is encrypted i guess

**Comment by Jasi2169** [comment]
@rovo89  can you provide me the v89 testzip ?

i have tested v88.1 and v88 both do not work for me on my galaxy s7 sm-g930F android 7.0 stock

i even tried simplehook to change model no.,board,product etc of build.prop ,changes are not seen properly like it does on below os versions

the logs shows no error anyway

rather then using shared_prefs i directly passed string to see but still changes can not be seen in AIDA64 app https://play.google.com/store/apps/details?id=com.finalwire.aida64&hl=en ,whereas marshmallow and below show perfectly

even tried to flash permissive kernel nothing seems to work out

```
@Override
    public void handleLoadPackage(final XC_LoadPackage.LoadPackageParam lpparam) throws Throwable {

       bld(cls,"MODEL","model");

        bld(cls,"MANUFACTURER","manuf");

        bld(cls,"HARDWARE","hrdwre");

        bld(cls,"BOARD","board");

        bld(cls,"PRODUCT","product");
}
////////////////////
private void bld(ClassLoader cls,String k,String v){
        XposedHelpers.setStaticObjectField(XposedHelpers.findClass("android.os.Build", cls),k,v);
    }
```

let me know if you need any help tracing the issue

**Comment by rovo89** [comment]
> can you provide me the v89 testzip ?

No, it's outdated. Everything of that is in v88.1, I changed the version number to indicate that there were only minor fixes.

Some suggestions:

- Try adding some logging to your `bld()` function.
- Use `Build.class` instead of looking it up.
- Also, try to access the field directly after your change, then you know for sure that you can except the values to be seen.
- Finally, make sure that the app you're testing with is actuallly accessing those properties via those fields. It may also read the properties directly from the system.

**Comment by Jasi2169** [comment]
kernel permissive or enforcing no issue

- main culprit 

xposed module can not read the preferences for me ,i have already format /data partition to make it mountable and remove encryption ,so the main issue is xposed can not read preferences and now can not used values saved from the app

**WORKAROUND

needed to change the package directory permission to 755 and woops done**

rovo can't xposed does this itself ?

like on marshmallow we needed to make the shared_prefs.xml file to word readable ourselves it was not possible by xposed.makeworldreadable call

like this we need to change worldreadable to /data/data/package directory to world readable
will xposed will support this in future ?
xposed.makeworldreadabledirectory something like this ?

also fix
xposed.makeworldreadable for file as well on marshmallow done by xposed itself ?

**Comment by Jeffdow85** [comment]
@rovo89 here are some logs (hope I've gotten the right stuff you need)  from Samsung Galaxy Note 8 international version touchwiz 7.1.1 it does boot fine and even functions fine when the phone first boots but intermittently hangs then soft reboots and takes ages to finish boot again. Any help or work arounds welcomed from others too. Thanks in advance. I feel like an iPhone user without xposed now that I've been spoilt with it for so long haha 😂 😂 

https://drive.google.com/file/d/0B5FsAtlFpGuvMmoxWjd6TWpIc2M/view?usp=drivesdk

**Comment by wanam** [comment]
@Jeffdow85 
Zygote is restarting every 10 seconds, same behavior as this one https://github.com/rovo89/XposedBridge/issues/205.

It's caused by Knox ASLR feature, as discussed with @rovo89 this may be disabled on the future updates.

**Comment by Jeffdow85** [comment]
@wanam Thanks for quick reply. Great to see you on board this project BTW Will go read now and exercise patience lol!

**Comment by rovo89** [comment]
I think the remaining issues should be fixed with v88.2.
